Add diagnostic to V2C

dgomes opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Adds the diagnostic platform to V2C

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

dgomes committed:
add diagnostic platform

joostlek commented on tests/components/v2c/snapshots/test_diagnostics.ambr:
```diff
@@ -0,0 +1,47 @@
+      'unique_id': 'ABC123',
+      'version': 1,
+    }),
+    'data': "<MagicMock name='Trydan().data' id='4511132720'>",
```
Something isn't properly mocked

joostlek commented on tests/components/v2c/test_diagnostics.py:
```diff
@@ -0,0 +1,44 @@
+from tests.components.diagnostics import get_diagnostics_for_config_entry
+from tests.typing import ClientSessionGenerator
+
+# Fields to exclude from snapshot as they change each run
+TO_EXCLUDE = {
+    "id",
+    "device_id",
+    "via_device_id",
+    "last_updated",
+    "last_changed",
+    "last_reported",
+}
```
These values aren't in the diagnostics test right?

dgomes replied:
based myself on existing tests and assumed to much 😓 

<removed>

joostlek commented on homeassistant/components/v2c/diagnostics.py:
```diff
@@ -0,0 +1,42 @@
+    coordinator_data = coordinator.evse.data
+    evse_raw_data = coordinator.evse.raw_data
+
+    diagnostic_data: dict[str, Any] = {
+        "config_entry": async_redact_data(entry.as_dict(), TO_REDACT),
+        "data": str(coordinator_data),
+        "raw_data": json_loads(evse_raw_data["content"]),  # type: ignore[arg-type]
+        "host_status": evse_raw_data["status_code"],
+    }
+
+    return diagnostic_data
```
```suggestion
    return {
        "config_entry": async_redact_data(entry.as_dict(), TO_REDACT),
        "data": str(coordinator_data),
        "raw_data": json_loads(evse_raw_data["content"]),  # type: ignore[arg-type]
        "host_status": evse_raw_data["status_code"],
    }
```

dgomes committed:
simplify return

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

dgomes committed:
address review comments

joostlek commented on homeassistant/components/v2c/diagnostics.py:
```diff
@@ -0,0 +1,40 @@
+    if TYPE_CHECKING:
+        assert coordinator.evse
+
+    coordinator_data = await coordinator.evse.get_data()
```
The coordinator data is already present in `coordinator.data` right?

dgomes replied:
It is, but after the failed mock (previous review commend) I realised this is a better solution - it will force a refresh which is actually more useful during debug

joostlek replied:
Sure but then you are not creating a diagnostic from the current state. If something goes wrong at a given moment and you want to capture that, snapshotting would refresh the lib and you would get new data that maybe doesn't have the same issue.

dgomes replied:
The important aspect is to capture the differences between the `raw_data` and the `TrydanData` 

The coordinator also updates quite often, it will be very difficult to time a diagnostic download with an ephemeral error.

But if necessary I can change back

joostlek replied:
But with what is raw_data filled then?

dgomes replied:
raw_data is literally the JSON coming out of the EVSE (the physical device)

The objective is to detect whenever the manufacturer changes the JSON and be able to easily provide an update to the integration

Quentame replied:
I agree with @joostlek : we prefer current coordinator data than a fresh one.

Plus, if API changes you will still see it 😌

dgomes replied:
reverted

Quentame commented on homeassistant/components/v2c/diagnostics.py:
```diff
@@ -0,0 +1,40 @@
+from .const import DOMAIN
+from .coordinator import V2CUpdateCoordinator
+
+CONF_TITLE = "title"
```
Not used ?

dgomes committed:
address review

dgomes committed:
use current data instead of getting refreshed data

emontnemery commented on homeassistant/components/v2c/manifest.json:
```diff
@@ -5,5 +5,5 @@
   "documentation": "https://www.home-assistant.io/integrations/v2c",
   "iot_class": "local_polling",
-  "requirements": ["pytrydan==0.6.1"]
+  "requirements": ["pytrydan==0.7.0"]
```
Can we bump the library in a separate PR?

dgomes replied:
https://github.com/home-assistant/core/pull/118726

bdraco committed:
Migrate recorder instance to use HassKey (#118673)

bdraco committed:
Convert mqtt to use a timer instead of task sleep loop (#118666)

mkmer committed:
Remove unintended translation key from blink (#118712)

dgomes committed:
Address reviews comments in #117147 (#118714)

joostlek committed:
Configure device in airgradient config flow (#118699)

bramkragten committed:
Update frontend to 20240603.0 (#118736)

dknowles2 committed:
Bump pydrawise to 2024.6.2 (#118608)

balloob committed:
Remove dispatcher from Tag entity (#118671)

* Remove dispatcher from Tag entity

* type

* Don't use  helper

* Del is faster than pop

* Use id in update

---------

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

rikroe committed:
Use snapshot_platform helper for BMW tests (#118735)

* Use snapshot_platform helper

* Remove comments

---------

Co-authored-by: Richard <rikroe@users.noreply.github.com>

tronikos committed:
Use ISO format when passing date to LLMs (#118705)

bdraco committed:
Small speed up to read-only database sessions (#118674)

synesthesiam committed:
Resolve areas/floors to ids in intent_script (#118734)

jbouwh committed:
Migrate Intergas InComfort/Intouch Lan2RF gateway  YAML to config flow (#118642)

* Add config flow

* Make sure the device is polled - refactor

* Fix

* Add tests config flow

* Update test requirements

* Ensure dispatcher has a unique signal per heater

* Followup on review

* Follow up comments

* One more docstr

* Make specific try blocks and refactoring

* Handle import exceptions

* Restore removed lines

* Move initial heater update in try block

* Raise issue failed import

* Update test codeowners

* Remove entity device info

* Remove entity device info

* Appy suggestions from code review

* Remove broad exception handling from entry setup

* Test coverage

joostlek committed:
Require firmware version 3.1.1 for airgradient (#118744)

jbouwh committed:
Add device info to incomfort entities (#118741)

* Add device info to incomfort entities

* Add DOMAIN import

zxdavb committed:
Harden evohome against failures to retrieve zone schedules (#118517)

balloob committed:
Update OpenAI prompt on each interaction (#118747)

balloob committed:
Google Gen AI: Copy messages to avoid changing the trace data (#118745)

jbouwh committed:
Revert "Allow MQTT device based auto discovery" (#118746)

Revert "Allow MQTT device based auto discovery (#109030)"

This reverts commit 585892f0678dc054819eb5a0a375077cd9b604b8.

jbouwh committed:
Do not log mqtt origin info if the log level does not allow it (#118752)

bdraco committed:
Speed up statistics_during_period websocket api (#118672)

synesthesiam committed:
Clean up exposed domains (#118753)

* Remove lock and script

* Add media player

* Fix tests

synesthesiam committed:
Bump intents to 2024.6.3 (#118748)

jbouwh committed:
Recover mqtt abbrevations optimizations (#118762)

Co-authored-by: J. Nick Koston <nick@koston.org>

ludeeus committed:
Update hass-nabucasa to version 0.81.1 (#118768)

bdraco committed:
Add os.walk to asyncio loop blocking detection (#118769)

bachya committed:
Don't require code to arm SimpliSafe (#118759)

starkillerOG committed:
Bump reolink-aio to 0.9.1 (#118655)

Co-authored-by: J. Nick Koston <nick@koston.org>

rikroe committed:
Allow per-sensor unit conversion on BMW sensors (#110272)

* Update BMW sensors to use device_class

* Test adjustments

* Trigger CI

* Remove unneeded cast

* Set suggested_display_precision to 0

* Rebase for climate_status

* Change charging_status to ENUM device class

* Add test for Enum translations

* Pin Enum sensor values

* Use snapshot_platform helper

* Remove translation tests

* Formatting

* Remove comment

* Use const.STATE_UNKOWN

* Fix typo

* Update strings

* Loop through Enum sensors

* Revert enum sensor changes

---------

Co-authored-by: Richard <rikroe@users.noreply.github.com>

joostlek committed:
Initial cleanup for Aladdin connect (#118777)

joostlek committed:
Use model from Aladdin Connect lib (#118778)

* Use model from Aladdin Connect lib

* Fix

epenet committed:
Add type hints for FixtureRequest in tests (#118779)

Commit:
Add ista EcoTrend integration (#118360)

* Add ista EcoTrend integration

* move code out of try

* Use account owners name as entry title

* update config flow tests

* add tests for init

* Add reauth flow

* Add tests for sensors

* add translations for reauth

* trigger statistics import on first refresh

* Move statistics and reauth flow to other PR

* Fix tests

* some changes

* draft_final_final

* remove unnecessary icons

* changed tests

* move device_registry test to init

* add text selectors

timmo001 committed:
Azure DevOps build sensor attributes to new sensors (#114948)

* Setup for split

* Adjust to allow for None

* Create

* Add missing

* Fix datetime parsing in Azure DevOps sensor

* Remove definition id and name

These aren't needed and will never change

* Add tests for each sensor

* Add tests for edge cases

* Rename translations

* Update

* Use base sensor descriptions

* Remove

* Drop status

using this later for an event entity

* Switch to timestamp

* Switch to timestamp

* Merge

* Update snapshot

* Improvements from @joostlek

* Update homeassistant/components/azure_devops/sensor.py

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

joostlek committed:
Add coordinator to Aladdin Connect (#118781)

joostlek committed:
Bump airgradient to 0.4.3 (#118776)

alexandrecuer committed:
Remove duplicate code in emoncms (#118610)

* Remove duplicate & property extra_state_attributes

* Add methods _update_attributes and _update_value

* correction in _update_value

* Update homeassistant/components/emoncms/sensor.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

* Update homeassistant/components/emoncms/sensor.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

* Update homeassistant/components/emoncms/sensor.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

* Update homeassistant/components/emoncms/sensor.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

---------

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

dependabot[bot] committed:
Bump dawidd6/action-download-artifact from 3.1.4 to 4 (#118772)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

joostlek committed:
Re-enable sensor platform for Aladdin Connect (#118782)

emontnemery committed:
Deduplicate light services.yaml (#118738)

* Deduplicate light services.yaml

* Remove support for .-keys

joostlek committed:
Move Aladdin stale device removal to init module (#118784)

epenet committed:
Cleanup unused FixtureRequest in tests (#118780)

epenet committed:
Move entity_registry_enabled_by_default to decorator [q-z] (#118793)

joostlek committed:
Make Ruuvi a brand (#118786)

joostlek committed:
Make Weatherflow a brand (#118785)

mj23000 committed:
Bang & Olufsen add overlay/announce play_media functionality (#113434)

* Add overlay service

* Convert custom service to play_media announce

* Remove debugging

emontnemery commented on homeassistant/components/v2c/diagnostics.py:
```diff
@@ -0,0 +1,38 @@
+    hass: HomeAssistant, entry: ConfigEntry
+) -> dict[str, Any]:
+    """Return diagnostics for a config entry."""
+    coordinator: V2CUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
```
In a separate PR, migrate the integration to store the coordinator in entry.runtime_data instead

emontnemery commented on homeassistant/components/v2c/diagnostics.py:
```diff
@@ -0,0 +1,38 @@
+    return {
+        "config_entry": async_redact_data(entry.as_dict(), TO_REDACT),
+        "data": str(coordinator_data),
+        "raw_data": json_loads(evse_raw_data["content"]),  # type: ignore[arg-type]
```
Should we guard against the raw data not necessarily being valid JSON?

dgomes replied:
indeed it guards the `data` but not the `raw_data`

will change

emontnemery commented on tests/components/v2c/snapshots/test_diagnostics.ambr:
```diff
@@ -0,0 +1,47 @@
+      'pref_disable_new_entities': False,
+      'pref_disable_polling': False,
+      'source': 'user',
+      'title': 'EVSE 1.1.1.1',
```
Is the title composed of the host name? If so, should we redact it too?

dgomes committed:
review comments

dgomes commented:
Sorry to have mess-up the commit will reopen clean :(
